### PR TITLE
add min valid jwt payload to API docs for structured authn config

### PR DESCRIPTION
**Pull request by aramase**
#### What type of PR is this?
/kind documentation
/kind api-change
(for the godoc update)

#### What this PR does / why we need it:
- Adds the minimum valid JWT payload for structured authentication configuration.
- Added a unit test to assert the min valid payload.

#### Which issue(s) this PR fixes:

part of https://github.com/kubernetes/kubernetes/issues/123318

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
[KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3331-structured-authentication-configuration
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go [hunk 1]
```diff
@@ -176,6 +176,12 @@ type AuthenticationConfiguration struct {
 	// authenticators is neither defined nor stable across releases.  Since
 	// each JWT authenticator must have a unique issuer URL, at most one
 	// JWT authenticator will attempt to cryptographically validate the token.
+	//
+	// The minimum valid JWT payload must contain the following claims:
+	// - "iss" (issuer)
+	// - "sub" (subject)
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:182]:
```suggestion
	// - "aud" (audience)
```

`sub` isn't required right?  Just any claim that is the for the username?

File: staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go [hunk 2]
```diff
@@ -176,6 +176,12 @@ type AuthenticationConfiguration struct {
 	// authenticators is neither defined nor stable across releases.  Since
 	// each JWT authenticator must have a unique issuer URL, at most one
 	// JWT authenticator will attempt to cryptographically validate the token.
+	//
+	// The minimum valid JWT payload must contain the following claims:
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:180]: I would format the claims as a JSON payload with example data.

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc_test.go [hunk 3]
```diff
@@ -2953,6 +2953,38 @@ func TestToken(t *testing.T) {
 			}`, valid.Unix()),
 			want: &user.DefaultInfo{},
 		},
+		// test to assert the minimum valid jwt payload
+		// the required claims are iss, aud, exp and <claimMappings.Username> (in this case sub).
```
Comment (enj) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc_test.go:2957]: I would use something other than `sub` in this test for the username claim.

File: test/integration/apiserver/oidc/oidc_test.go [hunk 4]
```diff
@@ -129,6 +129,7 @@ func runTests(t *testing.T, useAuthenticationConfig bool) {
 				oidcServer.TokenHandler().EXPECT().Token().Times(1).DoAndReturn(utilsoidc.TokenHandlerBehaviorReturningPredefinedJWT(
 					t,
 					signingPrivateKey,
+					// This asserts the minimum valid claims for an ID token required by the authenticator.
 					map[string]interface{}{
 						"iss": oidcServer.URL(),
 						"sub": defaultOIDCClaimedUsername,
```
Comment (enj) [inline test/integration/apiserver/oidc/oidc_test.go:135]: I would update this line and the associated config to use a different field for username.

**Review by enj — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go (hunk 1 above)
Comment (aramase) [reply staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:182]: yes, that's correct! should be fixed now

File: test/integration/apiserver/oidc/oidc_test.go [hunk 5]
```diff
@@ -941,6 +944,7 @@ func startTestAPIServerForOIDC[L utilsoidc.JosePublicKey](t *testing.T, oidcURL,
 			fmt.Sprintf("--oidc-client-id=%s", oidcClientID),
 			fmt.Sprintf("--oidc-ca-file=%s", oidcCAFilePath),
 			fmt.Sprintf("--oidc-username-prefix=%s", defaultOIDCUsernamePrefix),
+			"--oidc-username-claim=user",
```
Comment (enj) [inline test/integration/apiserver/oidc/oidc_test.go:947]: Hmm would be good to have at least one test that uses the default flag value.

Comment (aramase) [reply test/integration/apiserver/oidc/oidc_test.go]: Modified just the one test (`ID token is ok`) to validate with `user` claim and retained the rest of tests to use the `sub` claim as before.

**Pull request merged**